File: python/semantic_kernel/core_plugins/crew_ai/crew_ai_enterprise.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio
import logging
from typing import Any

import aiohttp
from pydantic import Field, ValidationError

from semantic_kernel.core_plugins.crew_ai.crew_ai_enterprise_client import CrewAIEnterpriseClient
from semantic_kernel.core_plugins.crew_ai.crew_ai_models import CrewAIEnterpriseKickoffState, CrewAIStatusResponse
from semantic_kernel.core_plugins.crew_ai.crew_ai_settings import CrewAISettings
from semantic_kernel.exceptions.function_exceptions import (
    FunctionExecutionException,
    FunctionResultError,
    PluginInitializationError,
)
from semantic_kernel.functions import kernel_function
from semantic_kernel.functions.kernel_function_from_method import KernelFunctionFromMethod
from semantic_kernel.functions.kernel_parameter_metadata import KernelParameterMetadata
from semantic_kernel.functions.kernel_plugin import KernelPlugin
from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.feature_stage_decorator import experimental

logger: logging.Logger = logging.getLogger(__name__)


@experimental
class CrewAIEnterprise(KernelBaseModel):
    """Class to interface with Crew.AI Crews from Semantic Kernel.

    This object can be used directly or as a plugin in the Kernel.
    """

    client: CrewAIEnterpriseClient
    polling_interval: float = Field(default=1.0)
    polling_timeout: float = Field(default=30.0)

    def __init__(
        self,
        endpoint: str | None = None,
        auth_token: str | None = None,
        polling_interval: float | None = 1.0,
        polling_timeout: float | None = 30.0,
        session: aiohttp.ClientSession | None = None,
        env_file_path: str | None = None,
        env_file_encoding: str | None = None,
    ):
        """Initialize a new instance of the class.  This object can be used directly or as a plugin in the Kernel.

        Args:
            endpoint (str | None, optional): The API endpoint.
            auth_token (str | None, optional): The authentication token.
            polling_interval (float, optional): The polling interval in seconds. Defaults to 1.0.
            polling_timeout (float, optional): The polling timeout in seconds. Defaults to 30.0.
            session (aiohttp.ClientSession | None, optional): The HTTP client session. Defaults to None.
            env_file_path (str | None): Use the environment settings file as a
                fallback to environment variables. (Optional)
            env_file_encoding (str | None): The encoding of the environment settings file. (Optional)
        """
        try:
            settings = CrewAISettings(
                endpoint=endpoint,
                auth_token=auth_token,
                polling_interval=polling_interval,
                polling_timeout=polling_timeout,
                env_file_path=env_file_path,
                env_file_encoding=env_file_encoding,
            )
        except ValidationError as ex:
            raise PluginInitializationError("Failed to initialize CrewAI settings.") from ex

        client = CrewAIEnterpriseClient(
            endpoint=settings.endpoint, auth_token=settings.auth_token.get_secret_value(), session=session
        )

        super().__init__(
            client=client,
            polling_interval=settings.polling_interval,
            polling_timeout=settings.polling_timeout,
        )

    async def kickoff(
        self,
        inputs: dict[str, Any] | None = None,
        task_webhook_url: str | None = None,
        step_webhook_url: str | None = None,
        crew_webhook_url: str | None = None,
    ) -> str:
        """Kickoff a new Crew AI task.

        Args:
            inputs (dict[str, Any], optional): The inputs for the task. Defaults to None.
            task_webhook_url (str | None, optional): The webhook URL for task updates. Defaults to None.
            step_webhook_url (str | None, optional): The webhook URL for step updates. Defaults to None.
            crew_webhook_url (str | None, optional): The webhook URL for crew updates. Defaults to None.

        Returns:
            str: The ID of the kickoff response.
        """
        try:
            kickoff_response = await self.client.kickoff(inputs, task_webhook_url, step_webhook_url, crew_webhook_url)
            logger.info(f"CrewAI Crew kicked off with Id: {kickoff_response.kickoff_id}")
            return kickoff_response.kickoff_id
        except Exception as ex:
            raise FunctionExecutionException("Failed to kickoff CrewAI Crew.") from ex

    @kernel_function(description="Get the status of a Crew AI kickoff.")
    async def get_crew_kickoff_status(self, kickoff_id: str) -> CrewAIStatusResponse:
        """Get the status of a Crew AI task.

        Args:
            kickoff_id (str): The ID of the kickoff response.

        Returns:
            CrewAIStatusResponse: The status response of the task.
        """
        try:
            status_response = await self.client.get_status(kickoff_id)
            logger.info(f"CrewAI Crew status for kickoff Id: {kickoff_id} is {status_response.state}")
            return status_response
        except Exception as ex:
            raise FunctionExecutionException(
                f"Failed to get status of CrewAI Crew with kickoff Id: {kickoff_id}."
            ) from ex

    @kernel_function(description="Wait for the completion of a Crew AI kickoff.")
    async def wait_for_crew_completion(self, kickoff_id: str) -> str:
        """Wait for the completion of a Crew AI task.

        Args:
            kickoff_id (str): The ID of the kickoff response.

        Returns:
            str: The result of the task.

        Raises:
            FunctionExecutionException: If the task fails or an error occurs while waiting for completion.
        """
        status_response: CrewAIStatusResponse | None = None
        state: str = CrewAIEnterpriseKickoffState.Pending

        async def poll_status():
            nonlocal state, status_response
            while state not in [
                CrewAIEnterpriseKickoffState.Failed,
                CrewAIEnterpriseKickoffState.Failure,
                CrewAIEnterpriseKickoffState.Success,
                CrewAIEnterpriseKickoffState.Not_Found,
            ]:
                logger.debug(
                    f"Waiting for CrewAI Crew with kickoff Id: {kickoff_id} to complete. Current state: {state}"
                )

                await asyncio.sleep(self.polling_interval)

                try:
                    status_response = await self.client.get_status(kickoff_id)
                    state = status_response.state
                except Exception as ex:
                    raise FunctionExecutionException(
                        f"Failed to wait for completion of CrewAI Crew with kickoff Id: {kickoff_id}."
                    ) from ex

        await asyncio.wait_for(poll_status(), timeout=self.polling_timeout)

        logger.info(f"CrewAI Crew with kickoff Id: {kickoff_id} completed with status: {state}")
        result = status_response.result if status_response is not None and status_response.result is not None else ""

        if state in ["Failed", "Failure"]:
            raise FunctionResultError(f"CrewAI Crew failed with error: {result}")

        return result

    def create_kernel_plugin(
        self,
        name: str,
        description: str,
        parameters: list[KernelParameterMetadata] | None = None,
        task_webhook_url: str | None = None,
        step_webhook_url: str | None = None,
        crew_webhook_url: str | None = None,
    ) -> KernelPlugin:
        """Creates a kernel plugin that can be used to invoke the CrewAI Crew.

        Args:
            name (str): The name of the kernel plugin.
            description (str): The description of the kernel plugin.
            parameters (List[KernelParameterMetadata] | None, optional): The definitions of the Crew's
            required inputs. Defaults to None.
            task_webhook_url (Optional[str], optional): The task level webhook URL. Defaults to None.
            step_webhook_url (Optional[str], optional): The step level webhook URL. Defaults to None.
            crew_webhook_url (Optional[str], optional): The crew level webhook URL. Defaults to None.

        Returns:
            dict[str, Any]: A dictionary representing the kernel plugin.
        """

        @kernel_function(description="Kickoff the CrewAI task.")
        async def kickoff(**kwargs: Any) -> str:
            args = self._build_arguments(parameters, kwargs)
            return await self.kickoff(
                inputs=args,
                task_webhook_url=task_webhook_url,
                step_webhook_url=step_webhook_url,
                crew_webhook_url=crew_webhook_url,
            )

        @kernel_function(description="Kickoff the CrewAI task and wait for completion.")
        async def kickoff_and_wait(**kwargs: Any) -> str:
            args = self._build_arguments(parameters, kwargs)
            kickoff_id = await self.kickoff(
                inputs=args,
                task_webhook_url=task_webhook_url,
                step_webhook_url=step_webhook_url,
                crew_webhook_url=crew_webhook_url,
            )
            return await self.wait_for_crew_completion(kickoff_id)

        return KernelPlugin(
            name,
            description,
            {
                "kickoff": KernelFunctionFromMethod(kickoff, stream_method=None, parameters=parameters),
                "kickoff_and_wait": KernelFunctionFromMethod(
                    kickoff_and_wait, stream_method=None, parameters=parameters
                ),
                "get_status": self.get_crew_kickoff_status,
                "wait_for_completion": self.wait_for_crew_completion,
            },
        )

    def _build_arguments(
        self, parameters: list[KernelParameterMetadata] | None, arguments: dict[str, Any]
    ) -> dict[str, Any]:
        """Builds the arguments for the CrewAI task from the provided parameters and arguments.

        Args:
            parameters (List[KernelParameterMetadata] | None): The metadata for the inputs.
            arguments (dict[str, Any]): The provided arguments.

        Returns:
            dict[str, Any]: The built arguments.
        """
        args = {}
        if parameters:
            for input in parameters:
                name = input.name
                if name not in arguments:
                    raise PluginInitializationError(f"Missing required input '{name}' for CrewAI.")
                args[name] = arguments[name]
        return args

    async def __aenter__(self):
        """Enter the session."""
        await self.client.__aenter__()
        return self

    async def __aexit__(self, *args, **kwargs):
        """Close the session."""
        await self.client.__aexit__()

File: dotnet/src/Plugins/Plugins.AI/CrewAI/CrewAIEnterprise.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Microsoft.SemanticKernel.Plugins.AI.CrewAI;

/// <summary>
/// A plugin for interacting with the a CrewAI Crew via the Enterprise APIs.
/// </summary>
public class CrewAIEnterprise
{
    private readonly ICrewAIEnterpriseClient _crewClient;
    private readonly ILogger _logger;
    private readonly TimeSpan _pollingInterval;

    /// <summary>
    /// The name of the kickoff function.
    /// </summary>
    public const string KickoffFunctionName = "KickoffCrew";

    /// <summary>
    /// The name of the kickoff and wait function.
    /// </summary>
    public const string KickoffAndWaitFunctionName = "KickoffAndWait";

    /// <summary>
    /// Initializes a new instance of the <see cref="CrewAIEnterprise"/> class.
    /// </summary>
    /// <param name="endpoint">The base URI of the CrewAI Crew</param>
    /// <param name="authTokenProvider"> Optional provider for auth token generation. </param>
    /// <param name="httpClientFactory">The HTTP client factory. </param>
    /// <param name="loggerFactory">The logger factory. </param>
    /// <param name="pollingInterval">Defines the delay time between status calls when pollin for a kickoff to complete.</param>
    public CrewAIEnterprise(Uri endpoint, Func<Task<string>> authTokenProvider, IHttpClientFactory? httpClientFactory = null, ILoggerFactory? loggerFactory = null, TimeSpan? pollingInterval = default)
    {
        Verify.NotNull(endpoint, nameof(endpoint));
        Verify.NotNull(authTokenProvider, nameof(authTokenProvider));

        this._crewClient = new CrewAIEnterpriseClient(endpoint, authTokenProvider, httpClientFactory);
        this._logger = loggerFactory?.CreateLogger(typeof(CrewAIEnterprise)) ?? NullLogger.Instance;
        this._pollingInterval = pollingInterval ?? TimeSpan.FromSeconds(1);
    }

    /// <summary>
    /// Internal constructor used for testing purposes.
    /// </summary>
    internal CrewAIEnterprise(ICrewAIEnterpriseClient crewClient, ILoggerFactory? loggerFactory = null)
    {
        Verify.NotNull(crewClient, nameof(crewClient));
        this._crewClient = crewClient;
        this._logger = loggerFactory?.CreateLogger(typeof(CrewAIEnterprise)) ?? NullLogger.Instance;
    }

    /// <summary>
    /// Kicks off (starts) a CrewAI Crew with the given inputs and callbacks.
    /// </summary>
    /// <param name="inputs">An object containing key value pairs matching the required inputs of the Crew.</param>
    /// <param name="taskWebhookUrl">The task level webhook Uri.</param>
    /// <param name="stepWebhookUrl">The step level webhook Uri.</param>
    /// <param name="crewWebhookUrl">The crew level webhook Uri.</param>
    /// <returns>The Id of the scheduled kickoff.</returns>
    /// <exception cref="KernelException"></exception>
    public async Task<string> KickoffAsync(
        object? inputs,
        Uri? taskWebhookUrl = null,
        Uri? stepWebhookUrl = null,
        Uri? crewWebhookUrl = null)
    {
        try
        {
            CrewAIKickoffResponse kickoffTask = await this._crewClient.KickoffAsync(
                inputs: inputs,
                taskWebhookUrl: taskWebhookUrl?.AbsoluteUri,
                stepWebhookUrl: stepWebhookUrl?.AbsoluteUri,
                crewWebhookUrl: crewWebhookUrl?.AbsoluteUri)
                .ConfigureAwait(false);

            this._logger.LogInformation("CrewAI Crew kicked off with Id: {KickoffId}", kickoffTask.KickoffId);
            return kickoffTask.KickoffId;
        }
        catch (Exception ex)
        {
            throw new KernelException(message: "Failed to kickoff CrewAI Crew.", innerException: ex);
        }
    }

    /// <summary>
    /// Gets the current status of the CrewAI Crew kickoff.
    /// </summary>
    /// <param name="kickoffId">The Id of the Crew kickoff.</param>
    /// <returns>A <see cref="CrewAIStatusResponse"/></returns>
    /// <exception cref="KernelException"></exception>"
    [KernelFunction]
    [Description("Gets the current status of the CrewAI Crew kickoff.")]
    public async Task<CrewAIStatusResponse> GetCrewKickoffStatusAsync([Description("The Id of the kickoff")] string kickoffId)
    {
        Verify.NotNullOrWhiteSpace(kickoffId, nameof(kickoffId));

        try
        {
            CrewAIStatusResponse statusResponse = await this._crewClient.GetStatusAsync(kickoffId).ConfigureAwait(false);

            this._logger.LogInformation("CrewAI Crew status for kickoff Id: {KickoffId} is {Status}", kickoffId, statusResponse.State);
            return statusResponse;
        }
        catch (Exception ex)
        {
            throw new KernelException(message: $"Failed to get status of CrewAI Crew with kickoff Id: {kickoffId}.", innerException: ex);
        }
    }

    /// <summary>
    /// Waits for the Crew kickoff to complete and returns the result.
    /// </summary>
    /// <param name="kickoffId">The Id of the crew kickoff.</param>
    /// <returns>The result of the Crew kickoff.</returns>
    /// <exception cref="KernelException"></exception>
    [KernelFunction]
    [Description("Waits for the Crew kickoff to complete and returns the result.")]
    public async Task<string> WaitForCrewCompletionAsync([Description("The Id of the kickoff")] string kickoffId)
    {
        Verify.NotNullOrWhiteSpace(kickoffId, nameof(kickoffId));

        try
        {
            CrewAIStatusResponse? statusResponse = null;
            var status = CrewAIKickoffState.Pending;
            do
            {
                this._logger.LogInformation("Waiting for CrewAI Crew with kickoff Id: {KickoffId} to complete. Current state: {Status}", kickoffId, status);
                await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
                statusResponse = await this._crewClient.GetStatusAsync(kickoffId).ConfigureAwait(false);
                status = statusResponse.State;
            }
            while (!this.IsTerminalState(status));

            this._logger.LogInformation("CrewAI Crew with kickoff Id: {KickoffId} completed with status: {Status}", kickoffId, statusResponse.State);

            return status switch
            {
                CrewAIKickoffState.Failed => throw new KernelException(message: $"CrewAI Crew failed with error: {statusResponse.Result}"),
                CrewAIKickoffState.Success => statusResponse.Result ?? string.Empty,
                _ => throw new KernelException(message: "Failed to parse unexpected response from CrewAI status response."),
            };
        }
        catch (Exception ex)
        {
            throw new KernelException(message: $"Failed to wait for completion of CrewAI Crew with kickoff Id: {kickoffId}.", innerException: ex);
        }
    }

    /// <summary>
    /// Creates a <see cref="KernelFunction"/> that can be used to invoke the CrewAI Crew.
    /// </summary>
    /// <param name="name">The name of the <see cref="KernelFunction"/></param>
    /// <param name="description">The description of the <see cref="KernelFunction"/></param>
    /// <param name="inputMetadata">The definitions of the Crew's required inputs.</param>
    /// <param name="taskWebhookUrl">The task level webhook Uri</param>
    /// <param name="stepWebhookUrl">The step level webhook Uri</param>
    /// <param name="crewWebhookUrl">The crew level webhook Uri</param>
    /// <returns>A <see cref="KernelFunction"/> that can invoke the Crew.</returns>
    /// <exception cref="KernelException"></exception>
    public KernelPlugin CreateKernelPlugin(
        string name,
        string description,
        IEnumerable<CrewAIInputMetadata>? inputMetadata,
        Uri? taskWebhookUrl = null,
        Uri? stepWebhookUrl = null,
        Uri? crewWebhookUrl = null)
    {
        var options = new KernelFunctionFromMethodOptions()
        {
            Parameters = inputMetadata?.Select(i => new KernelParameterMetadata(i.Name) { Description = i.Description, IsRequired = true, ParameterType = i.Type }) ?? [],
            ReturnParameter = new() { ParameterType = typeof(string) },
        };

        // Define the kernel function implementation for kickoff
        [KernelFunction(KickoffFunctionName)]
        [Description("kicks off the CrewAI Crew and returns the Id of the scheduled kickoff.")]
        async Task<string> KickoffAsync(KernelArguments arguments)
        {
            Dictionary<string, object?> args = BuildArguments(inputMetadata, arguments);

            return await this.KickoffAsync(
                inputs: args,
                taskWebhookUrl: taskWebhookUrl,
                stepWebhookUrl: stepWebhookUrl,
                crewWebhookUrl: crewWebhookUrl)
                .ConfigureAwait(false);
        }

        // Define the kernel function implementation for kickoff and wait for result
        [KernelFunction(KickoffAndWaitFunctionName)]
        [Description("kicks off the CrewAI Crew, waits for it to complete, and returns the result.")]
        async Task<string> KickoffAndWaitAsync(KernelArguments arguments)
        {
            Dictionary<string, object?> args = BuildArguments(inputMetadata, arguments);

            var kickoffId = await this.KickoffAsync(
                inputs: args,
                taskWebhookUrl: taskWebhookUrl,
                stepWebhookUrl: stepWebhookUrl,
                crewWebhookUrl: crewWebhookUrl)
                .ConfigureAwait(false);

            return await this.WaitForCrewCompletionAsync(kickoffId).ConfigureAwait(false);
        }

        return KernelPluginFactory.CreateFromFunctions(
            name,
            description,
            [
                KernelFunctionFactory.CreateFromMethod(KickoffAsync, new(), options),
                KernelFunctionFactory.CreateFromMethod(KickoffAndWaitAsync, new(), options),
                KernelFunctionFactory.CreateFromMethod(this.GetCrewKickoffStatusAsync),
                KernelFunctionFactory.CreateFromMethod(this.WaitForCrewCompletionAsync)
            ]);
    }

    #region Private Methods

    /// <summary>
    /// Determines if the Crew kikoff state is terminal.
    /// </summary>
    /// <param name="state">The state of the crew kickoff</param>
    /// <returns>A <see cref="bool"/> indicating if the state is a terminal state.</returns>
    private bool IsTerminalState(CrewAIKickoffState state)
    {
        return state == CrewAIKickoffState.Failed || state == CrewAIKickoffState.Failure || state == CrewAIKickoffState.Success || state == CrewAIKickoffState.NotFound;
    }

    private static Dictionary<string, object?> BuildArguments(IEnumerable<CrewAIInputMetadata>? inputMetadata, KernelArguments arguments)
    {
        // Extract the required arguments from the KernelArguments by name
        Dictionary<string, object?> args = [];
        if (inputMetadata is not null)
        {
            foreach (var input in inputMetadata)
            {
                // If a required argument is missing, throw an exception
                if (!arguments.TryGetValue(input.Name, out object? value) || value is null || value is not string strValue)
                {
                    throw new KernelException(message: $"Missing required input '{input.Name}' for CrewAI.");
                }

                // Since this KernelFunction does not have explicit parameters all the relevant inputs are passed as strings.
                // We need to convert the inputs to the expected types.
                if (input.Type == typeof(string))
                {
                    args.Add(input.Name, value);
                }
                else
                {
                    // Try to get a converter for the input type
                    var converter = TypeConverterFactory.GetTypeConverter(input.Type);
                    if (converter is not null)
                    {
                        args.Add(input.Name, converter.ConvertFrom(value));
                    }
                    else
                    {
                        // Try to deserialize the input as a JSON object
                        var objValue = JsonSerializer.Deserialize(strValue, input.Type);
                        args.Add(input.Name, objValue);
                    }
                }
            }
        }

        return args;
    }

    #endregion
}
